### ISSUE TITLE
rnp_uid_is_valid apparenty too strict

### ISSUE THREAD

**Issue by kaie** [rnpgp/rnp#1510]
I was given a public key, which contains multiple user IDs.

For some of the non-revoked user IDs, rnp_uid_is_valid reports "not valid".
This is contrary to gpg's report, which doesn't complain about those user IDs.

I've looked at the packet dump, trying to understand RNP's decision.

I can see two potential reasons.

(1) A user ID signature has four subpackets that are marked as critical, 2, 9, 27, 30 - does RNP not understand any of those and therefore decides it cannot accept the signature?

(2) A user ID signature has a subpacket type 9. Signature creation time, plus key expiration time from this subpacket are in the past. Does RNP decide that the signature is invalid, because the key is apparently no longer valid?

This might be an incorrect decision. The signature issuer key (sub packet 16) refers to the primary key. And the primary key is set to "never expire". Does this mean the primary key was defined to expired, but afterwards, the primary key was changed to not expire?

Does RNP use the issuer key expiration time directly? If yes, this could be a bug. It seems RNP should check if there is a newer expiration time definition available for this issuer key ID? If there is newer information available, and the newer information is different, than RNP should use the fresh expiration, and could conclude that this user ID is still valid.

**Comment by kaie** [comment]
Related to Thunderbird bug https://bugzilla.mozilla.org/show_bug.cgi?id=1673993

**Comment by ni4** [comment]
Thanks for reporting and for investigations.
(1) RNP understand that subpackets, so this should not affect the issue.
(2) That's the reason of the issue - when checking user validity, we check whether available certification is valid and doesn't expire key.
However, I'm still not sure what would be the correct decision (and that's why originally coded it to be more strict).
RFC 4880 says at 5.2.3.3:
```
   Subpackets that appear in a certification self-signature
   apply to the user name, and subpackets that appear in the subkey
   self-signature apply to the subkey.  Lastly, subpackets on the
   direct-key signature apply to the entire key.
```
And afterwards
```
   Implementing software should interpret a self-signature's preference
   subpackets as narrowly as possible.  For example, suppose a key has
   two user names, Alice and Bob.  Suppose that Alice prefers the
   symmetric algorithm CAST5, and Bob prefers IDEA or TripleDES.  If the
   software locates this key via Alice's name, then the preferred
   algorithm is CAST5; if software locates the key via Bob's name, then
   the preferred algorithm is IDEA.  If the key is located by Key ID,
   the algorithm of the primary User ID of the key provides the
   preferred symmetric algorithm.
```

As per this logic it seems that if key was located via userid with key-expiring certification, then key is logically expired and that's why userid is marked as invalid (i.e. unusable). From other side this applies to the key, not to the userid.

**Comment by kaie** [comment]
What is usually done when a key is extended - will all user ID binding signatures get updated?

Apparently the software that created the example key didn't do that. It kept the user ID signature unchanged.

Can we conclude that the "key expires" information in the "user ID binding signature" is purely for information purposes - only used if no better information is available?

In the scenario of a user ID binding, we always have the related key available. It's always the primary key, right? So in this scenario, you don't have to do any lookups to other keys. You could directly look at the related primary key. And if the primary key is different, you could treat the key expiration date in the binding signature as "obsolete" information.

**Comment by kaie** [comment]
I wish we had a document that clearly explained the purpose of "key expiration" inside a signature (as opposed to "signature expiration", which is obvious to understand).

It seems difficult to decide what exactly we should do.

I just performed an experiment.
- set system date to 2021-01-01
- used gpg to create a key that expires in one day
- result key has subpacket 9 in the user id signature
- set system date to 2021-01-03 (2 days later)
- used gpg to change the expiration to never expire of both primary key and subkey
- result key no longer has a subpacket 9 !!!

**Comment by kaie** [comment]
I will ask the owner of the example key, what software they used to change the expiration date of their key.

**Comment by MartinX3** [comment]
@kaie In my case it was either Enigmail or GnuPG on Linux, but I hardly think it was Enigmail
I did it last year after evaluating if a 4k key is still strong enough.

**Comment by ni4** [comment]
> I wish we had a document that clearly explained the purpose of "key expiration" inside a signature (as opposed to "signature expiration", which is obvious to understand).
...
> * used gpg to change the expiration to never expire of both primary key and subkey
> * result key no longer has a subpacket 9 !!!

Yeah, that's another complication - if key never expires, it may be indicated by absence of subpacket 9, or subpacket 9 with zero value. However, in case if direct-key signature doesn't have subpacket, what that indicates - never-expiring key, or 'this sig doesn't care about expiriation, use userid certification'.

Probably this question worth some discussion in the openpgp working group, just to introduce some rules, based on practice, to make things easier to understand and implement.

**Comment by kaie** [comment]
So, if Martin used GnuPG to change the expiration - why was the user ID signature not updated to match the new expiration?

Martin's primary key says "don't expire", but the user ID binding signature still lists a key expiration time for the primary key ID.

**Comment by MartinX3** [comment]
Ah, I shouldn't answer you email and this ticket the same time. :)

So the TL;DR from me is
Created 2014
Should expire April 2020
Extended in Enigmail to 31.03.2025

Thunderbird and GnuPG are saying 31.03.2025 for the primary and every sub key.
I don't see a "never expire" anywhere. :)

**Comment by ni4** [comment]
> So, if Martin used GnuPG to change the expiration - why was the user ID signature not updated to match the new expiration?

Probably GnuPG's behaviour was changed at some point, I didn't investigate that deep (just tried on 2.2.27 to change key's expiry via `--edit-key`), or has some different behaviour if used via Enigmail.

**Comment by ni4** [comment]
> Maybe I'm confused about the low level structure. If I do --list-packets of Martin's key, the first packet lists "expires: 0". But if I use --list-keys gpg reports that the primary key expires in 2025. I don't understand that.

Oh, that should be GnuPG's output thing which is adjusted to v3 keys, which did have key expiration days in key packet itself.
v4 keys doesn't have key expiration in the key packet itself, it is stored only in signatures (direct-key or userid certification).

**Comment by kaie** [comment]
I did another experiment. Looking at the long list of many signatures was difficult to interpret. I used --edit key and minimize on Martin's key, and then looked at the packet dump of the minimized key.

For some of the user IDs, the key expiration time is listed as 10 years.
For some other user IDs, it is listed as 5 years.

Why is that? Does GnuPG offer to individually set the expiration of user IDs? Or was a buggy version of GnuPG used to change the expiration to 10 years, which failed to change all user IDs?

**Comment by kaie** [comment]
> > Maybe I'm confused about the low level structure. If I do --list-packets of Martin's key, the first packet lists "expires: 0". But if I use --list-keys gpg reports that the primary key expires in 2025. I don't understand that.
> 
> Oh, that should be GnuPG's output thing which is adjusted to v3 keys, which did have key expiration days in key packet itself.
> v4 keys doesn't have key expiration in the key packet itself, it is stored only in signatures (direct-key or userid certification).

For the minimized key, I see one signature packet for each user ID and for each subkey. I don't see a separate signature for the primary key.

I guess this means, gnupg decides about primary key expiration based on the expiration of the subkeys.

**Comment by kaie** [comment]
> I did another experiment. Looking at the long list of many signatures was difficult to interpret. I used --edit key and minimize on Martin's key, and then looked at the packet dump of the minimized key.
> 
> For some of the user IDs, the key expiration time is listed as 10 years.
> For some other user IDs, it is listed as 5 years.
> 
> Why is that? Does GnuPG offer to individually set the expiration of user IDs? Or was a buggy version of GnuPG used to change the expiration to 10 years, which failed to change all user IDs?

Regardless of the answer to this question:

Apparently, GnuPG treats all user IDs as usable (if not revoked), even if the user ID signature says the issuer key is already expired.

I guess this means we should ignore the key expiration information in the used ID signature.

**Comment by ni4** [comment]
> For the minimized key, I see one signature packet for each user ID and for each subkey. I don't see a separate signature for the primary key.

There is no signature, it's just output of the GnuPG (it's a part of key packet dump code, notice expiredate variable):
```
  expiredate = 0;		/* have to get it from the selfsignature */
  max_expiredate = 0;
  algorithm = iobuf_get_noeof (inp);
  pktlen--;
  if (is_v5)
    {
      pkbytes = read_32 (inp);
      pktlen -= 4;
    }
  else
    pkbytes = 0;

  if (list_mode)
    {
      es_fprintf (listfp, ":%s key packet:\n"
                  "\tversion %d, algo %d, created %lu, expires %lu",
                  pkttype == PKT_PUBLIC_KEY ? "public" :
                  pkttype == PKT_SECRET_KEY ? "secret" :
                  pkttype == PKT_PUBLIC_SUBKEY ? "public sub" :
                  pkttype == PKT_SECRET_SUBKEY ? "secret sub" : "??",
                  version, algorithm, timestamp, expiredate);
      if (is_v5)
        es_fprintf (listfp, ", pkbytes %u\n", pkbytes);
      else
        es_fprintf (listfp, "\n");
    }

```

**Comment by kaie** [comment]
> > Why is that? Does GnuPG offer to individually set the expiration of user IDs? Or was a buggy version of GnuPG used to change the expiration to 10 years, which failed to change all user IDs?
> 
> Probably it's not a bug, but was a some sort of feature at some point: https://lists.gnupg.org/pipermail/gnupg-users/2015-September/054430.html

That post describes that --edit-key can be used to select each subkey separately and change the user ID expiration time.

I cannot confirm that with current gnupg. I added a second user ID to a key. I marked first user ID as primary. Saved. Then used --edit-key again. Used "uid 2" to select the second user ID. Typed command "expire", and gnupg explains that I'm changing the expiration date of the key. (It doesn't talk about the expiration of a user ID.)

I set a new expiration date. Saved. Then --list-packets shows the new expiration in the signatures for both user IDs.

Based on that, I think Martin's key has an inconsistent structure. If GnuPG was used to modify that key, I think that means that GnuPG may create such inconsistent keys.

Because the key expiration seems to be independent of user IDs, I suggest that we ignore the key expiration information in used ID signatures.

**Comment by MartinX3** [comment]
> Based on that, I think Martin's key has an inconsistent structure. If GnuPG was used to modify that key, I think that means that GnuPG may create such inconsistent keys.

Can I repair my key somehow?

**Comment by ni4** [comment]
> Can I repair my key somehow? Or should I just use `--edit-key` and reupload it?

`gpg --edit-key`, followed by `expire` should fix the problem. However, it would be a good idea to backup your key first, and check packet structure afterwards :-)

**Comment by MartinX3** [comment]
I did
1. `gpg --edit-key "ID"`
2. `expire`
3. `4y`
4. `save`
5. Import into Thunderbird

Sadly only the same 2 of 5 identities are visible in Thunderbird
And only the primary key got a new expire date.
I will look how to edit the expire date of the subkeys and try it again. :)

**Comment by ni4** [comment]
Hm, it should be somehow specific to your key - I tried test key with multiple uids, with different key expiration times, with `--faked-system-time` and so on, and each time GnuPG correctly updated key expiry for all userids. I used GnuPG 2.2.27.

**Comment by MartinX3** [comment]
That's sad
I created it with gpg in 2014 as RSA 4096-bit key

**Comment by ni4** [comment]
@MartinX3 No worries, while fix for this issue didn't get into just released v0.15.1, we will include it into the next update.

Another possible way to solve this could be (if you have time and desire to dig into it):
- backup your secret key
- import key to some temporary GnuPG directory (`gpg --homedir .gpgtmp --import your-secret-key.asc`)
- `gpg --homedir .gpgtmp --edit-key keyname` to minimize key, and delete all userids except problematic one.
- set expiration
- see whether it is correctly set for your userid, if not - delete userid completely and add the exact match back
- export this key to a file (say, `your-key-userid-1.asc`), and check it back via `gpg --list-packets`.
- import your original key to some other temporary GnuPG directory (`gpg --homedir .gpgkey --import your-secret-key.asc`)
- import `your-key-userid-1.asc` to `.gpgkey` dir. This should tell something like `1 signature imported`.
- export secret key from `.gpgkey` directory to file, and check whether edited userid appeared in Thunderbird, if yes - repeat this for each userid.

**Comment by kaie** [comment]
The question is difficult to understand. Especially the last part "since those have valid self-signature, which expires key" I don't understand.

**Comment by kaie** [comment]
In my opinion, the API rnp_uid_is_valid should give information about the user ID indepdently of global key validity.

**Comment by kaie** [comment]
Yes, Thunderbird checks for the global key's validity status, expiration, revocation. If the key is expired or revoked, we still want to show the key in the user interface. If a key is expired, we want to show the key, because the user might want to extend the expiration date. If the key is revoked, and it is a secret key, then we also want to continue to show the key - it might be a key that the user still needs to decrypt old email. And if we don't show the email addresses for the expired/revoked key, it is confusing.

So, in my opinion, rnp_uid_is_valid should say "valid", if the user ID by itself looks valid - even if the global key is expired/revoked.

**Comment by ni4** [comment]
> The question is difficult to understand. Especially the last part "since those have valid self-signature, which expires key" I don't understand.

Sorry that I wasn't clear. Under "valid self-signature, which expires key" I meant valid userid certification, made by the primary key, which has key expiration time which is in the past. I.e. in this case, if there is just one userid with this certification, key would be considered as expired, but userid would be valid.

Thanks for later replies - those clear my doubts.

**Comment by ni4** [comment]
Fixed via PR #1531

**Comment by MartinX3** [comment]
@kaie @ni4 could we reopen the issue ticket?
 I did `clean` my gpg key with `gpg (GnuPG) 2.2.32` and lost one of my uid's. in thunderbird.
 
 It's my gmail uid.
 https://keys.openpgp.org/search?q=130A7C7B515BECE7
